pkg/capabilities/validate.go

```diff
@@ -0,0 +1,210 @@
+package capabilities
+
+import (
+	"encoding/json"
+	"errors"
+	"reflect"
+
+	"github.com/invopop/jsonschema"
+	"github.com/mitchellh/mapstructure"
+	jsonvalidate "github.com/santhosh-tekuri/jsonschema/v5"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/values"
+)
+
+type Validator[Config any, Input any, Output any] struct {
+	ValidatorArgs
+
+	configType Config
+	inputType  Input
+	outputType Output
+
+	schemas map[string]string
+}
+
+var uriPrefix = "https://github.com/smartcontractkit/chainlink/capabilities/"
+
+var _ Validatable = (*Validator[any, any, any])(nil)
+
+type ValidatorArgs struct {
+	Info CapabilityInfo
+
+	// You can customize each one of the reflectors
+	// or leave them nil to use the default reflector.
+	//
+	// You can also override the default reflector by setting
+	// the DefaultReflector field.
+	DefaultReflector *jsonschema.Reflector
+
+	SchemaReflector *jsonschema.Reflector
+	ConfigReflector *jsonschema.Reflector
+	InputReflector  *jsonschema.Reflector
+	OutputReflector *jsonschema.Reflector
+}
+
+func NewValidator[Config any, Input any, Output any](args ValidatorArgs) Validator[Config, Input, Output] {
+	baseID := jsonschema.ID(uriPrefix)
+	defaultReflector := &jsonschema.Reflector{ExpandedStruct: true, DoNotReference: true,
+		BaseSchemaID: baseID,
+	}
+	if args.DefaultReflector != nil {
+		defaultReflector = args.DefaultReflector
+	}
+
+	if args.SchemaReflector == nil {
+		args.SchemaReflector = defaultReflector
+	}
+
+	if args.ConfigReflector == nil {
+		args.ConfigReflector = defaultReflector
+	}
+
+	if args.InputReflector == nil {
+		args.InputReflector = defaultReflector
+	}
+
+	if args.OutputReflector == nil {
+		args.OutputReflector = defaultReflector
+	}
+
+	return Validator[Config, Input, Output]{
+		ValidatorArgs: args,
+		schemas:       make(map[string]string),
+	}
+}
+
+func (v *Validator[Config, Input, Output]) Schema() (string, error) {
+	type Combined struct {
```

Comment:
This (and CombinedWithoutInput below) can be private since they're declared inside a function.